### PR TITLE
chore: fix pty-max-limit flake

### PR DESCRIPTION
**Pull request by evgeniy-scherbina**
# Fix PTY exhaustion flake in parallel tests

## Problem

Tests were flaking due to PTY exhaustion when running in parallel, particularly on macOS. The issue manifested in `TestNewServer_CloseActiveConnections/Shutdown` where multiple parallel test packages would exhaust the system's PTY limit (`kern.tty.ptmx_max` on macOS, typically 511), causing tests to fail with PTY creation errors.

## Solution

Implemented a semaphore-based PTY limit that restricts the number of concurrent PTYs created during test execution. The limit is calculated as `PTY_MAX_TOTAL / TEST_NUM_PARALLEL_PACKAGES`, ensuring each test package gets a fair share of the system's PTY resources.

### Implementation Details

1. **Semaphore in `pty` package** (`pty/pty.go`):
   - Semaphore is initialized in `init()` when `testing.Testing()` is true
   - Capacity is calculated from `PTY_MAX_TOTAL` (default: 4096) divided by `TEST_NUM_PARALLEL_PACKAGES` (default: 8)
   - Semaphore slots are acquired before PTY creation and released when the PTY is closed
   - Production builds never set the semaphore (no overhead)

2. **CI configuration** (`.github/actions/test-go-pg/action.yaml`):
   - New step "Set PTY limit for tests" reads the system PTY limit:
     - macOS: `sysctl -n kern.tty.ptmx_max` (e.g., 511)
     - Linux: `/proc/sys/kernel/pty/max` (e.g., 4096)
   - Value is validated as numeric before writing to `GITHUB_ENV`
   - Step is skipped on Windows (not needed for ConPTY)

3. **PTY creation** (`pty/pty_other.go`):
   - `acquireTestSemaphore()` is called before `pty.Open()`
   - `releaseTestSemaphore()` is called in `Close()` when the PTY is closed
   - Semaphore slot is held for the lifetime of the PTY, ensuring we limit concurrent PTYs in use

4. **Windows** (`pty/pty_windows.go`):
   - Semaphore is not applied on Windows (ConPTY doesn't have the same exhaustion issues)

## Example

With `PTY_MAX_TOTAL=511` (macOS default) and `TEST_NUM_PARALLEL_PACKAGES=8`:
- Capacity per package = `511 / 8 = 63` PTYs
- Each test package can create up to 63 concurrent PTYs
- Total across all packages ≤ 511 (system limit)

## Related

Fixes flake in `TestNewServer_CloseActiveConnections/Shutdown` (https://github.com/coder/internal/issues/558)

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -35,7 +35,7 @@ jobs:
       tailnet-integration: ${{ steps.filter.outputs.tailnet-integration }}
     steps:
       - name: Harden Runner
-        uses: step-security/harden-runner@e3f713f2d8f53843e71c69a996d56f51aa9adfb9 # v2.14.1
+        uses: step-security/harden-runner@5ef0c079ce82195b2a36a210272d6b661572d83e # v2.14.2
```
Comment (zedkipp) [inline .github/workflows/ci.yaml:38]: Does this need to be bumped in this PR?

File: .github/actions/test-go-pg/action.yaml [hunk 2]
```diff
@@ -55,6 +55,29 @@ runs:
...
+        set -euo pipefail
+        # Capacity per package = PTY_MAX_TOTAL / TEST_NUM_PARALLEL_PACKAGES
+        pty_max=""
+        if [[ "$RUNNER_OS" == "macOS" ]]; then
+          pty_max=$(sysctl -n kern.tty.ptmx_max)
+        elif [[ "$RUNNER_OS" == "Linux" ]]; then
+          if [[ -f /proc/sys/kernel/pty/max ]]; then
+            pty_max=$(cat /proc/sys/kernel/pty/max)
+          else
+            pty_max="4096"
```
Comment (zedkipp) [inline .github/actions/test-go-pg/action.yaml:71]: Where does `4096` come from?

File: .github/actions/test-go-pg/action.yaml [hunk 3]
```diff
@@ -55,6 +55,29 @@ runs:
       run: |
         go run scripts/embedded-pg/main.go -path "${EMBEDDED_PG_PATH}" -cache "${EMBEDDED_PG_CACHE_DIR}"
 
+    - name: Set PTY limit for tests
+      if: runner.os != 'Windows'
+      shell: bash
+      run: | # zizmor: ignore[github-env] Value is validated as numeric before writing
+        set -euo pipefail
+        # Capacity per package = PTY_MAX_TOTAL / TEST_NUM_PARALLEL_PACKAGES
```
Comment (zedkipp) [inline .github/actions/test-go-pg/action.yaml:63]: I think this the calculation gives the _per-process_  allocation, no?

Comment (zedkipp) [inline .github/actions/test-go-pg/action.yaml:63]: Would be good to add a short sentence here explaining why this exists

File: pty/pty.go
Comment (zedkipp) [inline pty/pty.go:1]: Could this logic be extracted out of the production code and into the `pytest` package? It also seems like it could be much simpler if it wrapped the existing `newPty()` func. Maybe something like

```
  func New(t *testing.T, opts ...pty.Option) *PTY {
      t.Helper()

      acquireSemaphore(t) 

      ptty, err := newTestPTY(opts...)
      if err != nil {
          releaseSemaphore()
          require.NoError(t, err)
      }

      // ... existing setup ...

      t.Cleanup(func() {
          _ = r.Close()
          releaseSemaphore() 
      })
      return r
  }
```

Comment (spikecurtis) [reply pty/pty.go:1]: Dicey because ptys are used directly in other contexts, like the workspace agent for SSH sessions.

File: pty/pty.go [hunk 4]
```diff
@@ -1,17 +1,119 @@
...
+		capacity = 1
+	}
+	SetTestSemaphore(capacity)
+}
+
+// SetTestSemaphore configures a semaphore to limit concurrent PTY creation in tests.
+// It is called from pty's init() when testing.Testing() is true;
+// capacity is the maximum number of PTYs that can be created concurrently.
+// If capacity is 0 or negative, the semaphore is disabled.
+func SetTestSemaphore(capacity int) {
```
Comment (spikecurtis) [inline pty/pty.go:60]: I don't understand the reason this function is exported outside the package. If you kept it internal, then it would only be called from `init()` and you wouldn't need the `testSemaphoreMu`.

I don't really think it is safe to call outside `init()` anyway, since a goroutine might have already acquired the semaphore, so the release would be against a different channel.

**Review by spikecurtis — COMMENTED** [review]
The stated problem is that there is a _system-wide_ limit on the number of PTYs. I'm not convinced that setting a per-test-process limit on the PTYs is a good idea.

First of all, there will be more than just our Go tests running on most systems, and so other things outside our control could consume PTYs and cause tests to fail in spite of this system.

Secondly, it likely makes for inefficient use of PTYs. Packages that use a lot of PTYs will saturate at 1/8th the total available even if nothing else is using the remaining 7/8ths of the available resources.

Just conceptually, the Kernel is managing the PTY resources from a system-wide pool. We shouldn't also need to separately manage the same resources. If, as this PR is designed, we are willing to wait up to 30s to obtain a PTY if one is not immediately available, why not just retry with a short backoff (e.g. 1s)?

File: .github/actions/test-go-pg/action.yaml [hunk 5]
```diff
@@ -55,6 +55,29 @@ runs:
...
+        pty_max=""
+        if [[ "$RUNNER_OS" == "macOS" ]]; then
+          pty_max=$(sysctl -n kern.tty.ptmx_max)
```
Comment (mafredri) [inline .github/actions/test-go-pg/action.yaml:66]: Have we tried `sudo sysctl -w` instead to raise the limit? I think `999` is the max on macOS though.

File: pty/pty.go [hunk 6]
```diff
@@ -1,17 +1,119 @@
...
+		return
+	}
+	maxTotal := defaultPTYMaxTotal
+	if s := os.Getenv("PTY_MAX_TOTAL"); s != "" {
+		if n, err := strconv.Atoi(s); err == nil && n > 0 {
+			maxTotal = n
+		}
+	}
+	parallelPkgs := 8
+	if s := os.Getenv("TEST_NUM_PARALLEL_PACKAGES"); s != "" {
```
Comment (mafredri) [inline pty/pty.go:44]: This seems quite fragile, the env could easily be renamed or removed and then this logic is never invoked.

**Review by mafredri — COMMENTED** [review]
Considering the parallelism of our test-suite, I'd expect resource allocation issues is not limited to just PTYs. If we're running into this, we could instead consider upping limits, reducing parallelism, or going over which are the main problematic tests and remove parallelism from those (e.g. no `t.Parallel()`).

The main problems I see with this approach is adding test-code to a production package (have we considered build-flags?) and all the edge-cases in which it breaks (accurate detection of active parallelism, assumption that the rest of the system isn't under load, breaking expectations of typical context-timeouts in tests). We're also assuming each package is doing PTY testing which is unlikely to be the case. Without verification, I'd expect this number to be around 3-5 packages in total.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (evgeniy-scherbina) [reply .github/workflows/ci.yaml:38]: Yeah, this CI job is broken with `v2.14.1` — it just wasn’t triggered before.

File: .github/actions/test-go-pg/action.yaml (hunk 2 above)
Comment (evgeniy-scherbina) [reply .github/actions/test-go-pg/action.yaml:71]: It's default on dogfood and common default on Linux systems

File: .github/actions/test-go-pg/action.yaml (hunk 3 above)
Comment (evgeniy-scherbina) [reply .github/actions/test-go-pg/action.yaml:63]: The idea is `go test` launches `TEST_NUM_PARALLEL_PACKAGES` processes in parallel, so we give each process `PTY_MAX_TOTAL / TEST_NUM_PARALLEL_PACKAGES` PTYs.

I guess `per-process` / `per-package` kind of the same in this context?

File: pty/pty.go
Comment (evgeniy-scherbina) [reply pty/pty.go:1]: @zedkipp `newPty` in `pty_other.go` also used in `startPty` function, not only `New` function.
We MUST keep track of all PTY sessions, no matter how they are created.

**Comment by evgeniy-scherbina** [comment]
> The stated problem is that there is a _system-wide_ limit on the number of PTYs. I'm not convinced that setting a per-test-process limit on the PTYs is a good idea.
> 
> First of all, there will be more than just our Go tests running on most systems, and so other things outside our control could consume PTYs and cause tests to fail in spite of this system.
> 
> Secondly, it likely makes for inefficient use of PTYs. Packages that use a lot of PTYs will saturate at 1/8th the total available even if nothing else is using the remaining 7/8ths of the available resources.
> 
> Just conceptually, the Kernel is managing the PTY resources from a system-wide pool. We shouldn't also need to separately manage the same resources. If, as this PR is designed, we are willing to wait up to 30s to obtain a PTY if one is not immediately available, why not just retry with a short backoff (e.g. 1s)?

@spikecurtis This approach was originally introduced by you and @mafredri :
https://github.com/coder/internal/issues/558#issuecomment-2857260671
But I agree with both points.

`why not just retry with a short backoff (e.g. 1s)`- yeah, that's probably a better idea.

`UPD`: I see that there are a lot of fair critique, so going to close this PR and reopen new one which will rely on:
- retries proposed by Spike
- increasing limits if it's possible
?

File: pty/pty.go (hunk 6 above)
Comment (evgeniy-scherbina) [reply pty/pty.go:44]: It's introduced and used here:
https://github.com/coder/coder/blob/main/.github/actions/test-go-pg/action.yaml#L58-L77

I also tried different approach, getting value of `go test -p` flag, but I had issues with that, because we initialize semaphore in init function before `-p` value is available.

**Comment by evgeniy-scherbina** [comment]
Moved to https://github.com/coder/coder/pull/22147

**Comment by spikecurtis** [comment]
Sorry for being inconsistent. Clearly didn't think things through last spring. I think what solidified it for me was seeing that 30s timeout.
